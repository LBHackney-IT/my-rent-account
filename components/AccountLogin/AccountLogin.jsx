import React, { useState } from "react";
import axios from "axios";
import PropTypes from "prop-types";
import { useForm } from "react-hook-form";
import Router from "next/router";
import isPostcodeValid from "uk-postcode-validator";

import { Button, TextInput, Checkboxes } from "components/Form";
import ErrorMessage from "components/ErrorMessage/ErrorMessage";

const AccountLogin = ({
  adminDetails,
  onSubmit,
  onAsyncSubmit,
  submitText,
}) => {
  const [submitting, setSubmitting] = useState();
  const { register, errors, handleSubmit } = useForm();
  const [error, setError] = useState();
  const onLogin = async (params) => {
    try {
      setSubmitting(true);
      setError();
      const account = await axios.get("/api/accounts", { params });

      if (adminDetails && adminDetails.isAdmin) {
        const extraAdminAuditParams = {
          user: adminDetails.adminEmail,
        };
        await axios.post("/api/adminaudit", {
          ...params,
          ...extraAdminAuditParams,
        });
      } else {
        await axios.post("/api/audit", { ...params, ...account });
      }

      onSubmit && onSubmit(params);
      onAsyncSubmit && (await onAsyncSubmit(params));
      Router.push("/account");
    } catch (e) {
      console.error(e);
      setError(
        e.response.status === 404 ? (
          <div>
            <p>
              This rent account number and postcode combination is incorrect.
            </p>
            Please check if you have entered the correct information.
          </div>
        ) : (
          e.response.data
        )
      );
    }
    setSubmitting(false);
  };
  return (
    <form onSubmit={handleSubmit(onLogin)}>
      <TextInput
        label="Rent account number:"
        hint="You will find this on your rent card or statement"
        name="accountNumber"
        placeholder="e.g. 1234567890"
        inputClassName="govuk-input--width-10"
        inputMode="numeric"
        error={errors.accountNumber}
        register={register({
          required: "Rent account number is required",
          minLength: {
            value: 10,
            message: "You are missing some digits in the rent account number",
          },
          pattern: {
            value: /^[0-9]+$/,
            message: "Account Number must be a number",
          },
        })}
      />
      <TextInput
        label="Post code:"
        name="postcode"
        placeholder="e.g. E8 1EA"
        inputClassName="govuk-input--width-10"
        error={errors.postcode}
        register={register({
          required: "Post code is required",
          validate: (value) =>
            isPostcodeValid(value) || "You need a valid post code",
        })}
      />
      {adminDetails && adminDetails.isAdmin && (
        <Checkboxes
          options={[
            {
              value: "yes",
              label: "Simulate CSSO",
            },
          ]}
          name="simulateCSSO"
          register={register}
        />
      )}
      <Button text={submitText} disabled={submitting} />
      {error && <ErrorMessage text={error} />}
    </form>
  );
};

AccountLogin.propTypes = {
  adminDetails: PropTypes.object,
  onSubmit: PropTypes.func,
  onAsyncSubmit: PropTypes.func,
  submitText: PropTypes.string.isRequired,
};

export default AccountLogin;
