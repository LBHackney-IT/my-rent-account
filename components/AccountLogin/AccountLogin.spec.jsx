import React from "react";
import { render } from "@testing-library/react";

import AccountLogin from "./AccountLogin";

describe("AccountLogin component", () => {
  const props = {
    onSubmit: jest.fn(),
    submitText: "foo",
  };
  it("should render properly", () => {
    const { asFragment } = render(<AccountLogin {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
