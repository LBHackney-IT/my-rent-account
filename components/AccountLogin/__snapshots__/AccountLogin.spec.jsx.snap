// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountLogin component should render properly 1`] = `
<DocumentFragment>
  <form>
    <div
      class="govuk-form-group"
    >
      <label
        class="govuk-label govuk-label--m"
        for="accountNumber"
      >
        Rent Account Number:
      </label>
      <span
        class="govuk-hint"
        id="accountNumber-hint"
      >
        You will find this on your rent card or statement
      </span>
      <input
        aria-describedby="accountNumber-hint"
        class="govuk-input govuk-input--width-10"
        data-testid="accountNumber"
        id="accountNumber"
        inputmode="numeric"
        name="accountNumber"
        placeholder="e.g. 1234567890"
        type="text"
      />
    </div>
    <div
      class="govuk-form-group"
    >
      <label
        class="govuk-label govuk-label--m"
        for="postcode"
      >
        Post Code:
      </label>
      <input
        class="govuk-input govuk-input--width-10"
        data-testid="postcode"
        id="postcode"
        name="postcode"
        placeholder="e.g. E8 1EA"
        type="text"
      />
    </div>
    <button
      class="govuk-button"
      data-module="govuk-button"
    >
      foo
    </button>
  </form>
</DocumentFragment>
`;
